### PR TITLE
Some error bar clean ups

### PR DESCRIPTION
**Pull request by etpinard**
@alexcjohnson @mdtusz @cldougl @bpostlethwaite 

This PR:

- fixes default logic for error bar `type: 'sqrt'` 
- adds test for error bar `type: 'sqrt'` (about time)
- reuse svg error bar logic in 3d convert step
- split error bar `default`, `calc` and reusable compute-error-magnitude routine into separate files
- remove an used function

### CODE REVIEW

File: src/components/errorbars/compute_error.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * @param {numeric} dataPt data point from where to compute the error magnitue
+ * @param {number} index index of dataPt in its corresponding data array
+ * @param {object} opts error bar attributes
+ *
+ * @return {array of two numbers}
+ *      - error[0] : error magnitude in the negative direction
+ *      - error[1] : " " " " positive "
+ */
+module.exports = function computeError(dataPt, index, opts) {
```
Comment (etpinard) [inline src/components/errorbars/compute_error.js]: reusable error bar data/constant/sqrt/percent logic

File: src/components/errorbars/defaults.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        containerIn = traceIn[objName] || {};
+
+    function coerce (attr, dflt) {
+        return Lib.coerce(containerIn, containerOut, attributes, attr, dflt);
+    }
+
+    var hasErrorBars = (
+        containerIn.array !== undefined ||
+        containerIn.value !== undefined ||
+        containerIn.type === 'sqrt'
```
Comment (etpinard) [inline src/components/errorbars/defaults.js]: cc @alexcjohnson

File: src/components/errorbars/calc.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+
+'use strict';
+
+var isNumeric = require('fast-isnumeric');
+
+var Plots = require('../../plots/plots');
+var Axes = require('../../plots/cartesian/axes');
+
+var compureError = require('./compute_error');
```
Comment (mdtusz) [inline src/components/errorbars/calc.js]: `computeError` perhaps?

File: src/components/errorbars/calc.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            trace = calcTrace[0].trace;
+
+        if(!Plots.traceIs(trace, 'errorBarsOK')) continue;
+
+        var xObj = trace.error_x || {},
+            yObj = trace.error_y || {},
+            xa = Axes.getFromId(gd, trace.xaxis),
+            ya = Axes.getFromId(gd, trace.yaxis),
+            xVis = xObj.visible && ['linear', 'log'].indexOf(xa.type)!==-1,
+            yVis = yObj.visible && ['linear', 'log'].indexOf(ya.type)!==-1;
```
Comment (mdtusz) [inline src/components/errorbars/calc.js]: Spacing between `!==` and `1` for readability.

File: src/components/errorbars/calc.js (hunk 3 above)
Comment (etpinard) [inline src/components/errorbars/calc.js]: nice catch. Thanks

File: src/components/errorbars/calc.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            yVals = [];
+
+        for(var j = 0; j < calcTrace.length; j++) {
+            var calcPt = calcTrace[j],
+                calcX = calcPt.x,
+                calcY = calcPt.y;
+
+            if(!isNumeric(ya.c2l(calcY)) || !isNumeric(xa.c2l(calcX))) continue;
+
+            var errorY = compureError(calcY, j, yObj);
```
Comment (mdtusz) [inline src/components/errorbars/calc.js]: compureError here too

File: src/components/errorbars/calc.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            if(isNumeric(errorY[0]) && isNumeric(errorY[1])) {
+                calcPt.ys = calcY - errorY[0];
+                calcPt.yh = calcY + errorY[1];
+                yVals.push(calcPt.ys, calcPt.yh);
+            }
+
+            var errorX = compureError(calcX, j, xObj);
```
Comment (mdtusz) [inline src/components/errorbars/calc.js]: And here.

File: src/components/errorbars/compute_error.js (hunk 1 above)
Comment (alexcjohnson) [inline src/components/errorbars/compute_error.js]: Looks good to me. Don't need to do it now, but it occurs to me we could boost performance by just evaluating `opts` once and returning a more minimal function to evaluate in the loop:
```
function computeError(opts) {
    if(type === 'data') {
        if(opts.symmetric || opts.arrayminus === undefined) {
            var array = opts.array
            return function(dataPt, index) {
                var val = +(array[index]);
                return [val, val];
            }
        }
        // etc...
    }
}
```

Comment (etpinard) [inline src/components/errorbars/compute_error.js]: good call.

Comment (etpinard) [inline src/components/errorbars/compute_error.js]: all about the :racehorse:

**Comment by alexcjohnson** [comment]
Thanks @etpinard , :dancer: for my part.

File: src/components/errorbars/compute_error.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            arrayminus = opts.arrayminus;
+
+        if(symmetric || arrayminus === undefined) {
+            return function computeError(dataPt, index) {
+                var val = +(array[index]);
+                return [val, val];
+            };
+        }
+        else {
+            return function computeError(dataPt, index) {
```
Comment (etpinard) [inline src/components/errorbars/compute_error.js]: @bpostlethwaite @mdtusz better?

File: src/components/errorbars/compute_error.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            };
+        }
+    }
+    else {
+        var value = opts.value,
+            valueminus = opts.valueminus;
+
+        if(symmetric || valueminus === undefined) {
+            return function computeError(dataPt) {
+                var val = getErrorVal(type, dataPt, value);
```
Comment (alexcjohnson) [inline src/components/errorbars/compute_error.js]: Not to be pedantic, but you're still switching on type many times, you could do it once:
```
function getErrorFunction(type) {
    if(type === 'percent') return function(dataPt, value) {
        return Math.abs(dataPt * value / 100);
    }
    // etc
}
```
and then here:
```
var value = opts.value,
    valueminus = opts.valueminus,
    errorfunc = getErrorFunction(type);

if(symmetric || valueminus === undefined) {
    return function(dataPt) {
        var val = errorFunc(dataPt, value);
        return [val, val];
    }
}
else {
    return function(dataPt) {
        return [
            errorFunc(dataPt, value),
            errorFunc(dataPt, valueminus)
        ];
    }
}
```

Comment (etpinard) [inline src/components/errorbars/compute_error.js]: good point. all about the :racehorse: 

We have quite a few of these reusable calc functions, might as well set a good benchmark for them now.

File: src/components/errorbars/compute_error.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        };
+    }
+    if(type === 'constant') {
+        return function() {
+            return Math.abs(value);
+        };
+    }
+    if(type === 'sqrt') {
+        return function(dataPt) {
+            return Math.sqrt(Math.abs(dataPt));
```
Comment (etpinard) [inline src/components/errorbars/compute_error.js]: @alexcjohnson is this what you had in mind?

Comment (alexcjohnson) [inline src/components/errorbars/compute_error.js]: :heart_eyes:

Comment (mdtusz) [inline src/components/errorbars/compute_error.js]: Seems like an ideal place to use a `switch` statement.

Comment (etpinard) [inline src/components/errorbars/compute_error.js]: We've been reluctant to use `swtich` statements in plotly.js so far. Maybe this is the time to use them more. 

@bpostlethwaite `switch` statements are kinda messy in js, are they considered a good pattern?

**Pull request merged**